lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskRuntime.cs

```diff
@@ -173,6 +173,23 @@ public void Close(byte[] message)
                 _currentStatus.SetException(TaskClientCodeException.Create(
                     TaskId, ContextId, "Error during Close().", e));
             }
+            finally
+            {
+                Logger.Log(Level.Info, "TaskRuntime.Close, finally._userTask: " + _userTask);
```

Comment:
That's true. Ideally task should implement the proper close logic, that is called cooperative cancellation model  (https://msdn.microsoft.com/en-us/library/dd997364(v=vs.110).aspx). We already allow the task to add its own close handler. It is the task's responsibility to let the thread to return gracefully. 

Even if we add CancellationToken in the Call() method, it is still up to the client to respect it in the code. We have no way to control. It still falls to Cooperative cancellation model .
 
Checking if the thread is still Alive and Abort it might be a way, but Abort() is not recommended approach. 

In the current logic, at least we disposed object to release the resource. If the thread is still running, and find the object is disposed, it should throw exception and it will reach to the catch inside StartTaskOnNewThread in TaskRuntime. 
If you have any better solution, please share it.